builtin/logical/totp/path_keys.go

```
@@ -0,0 +1,399 @@
+	entry, err := s.Get("key/" + n)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var result keyEntry
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func (b *backend) pathKeyDelete(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	err := req.Storage.Delete("key/" + data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (b *backend) pathKeyRead(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	key, err := b.Key(req.Storage, data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+	if key == nil {
+		return nil, nil
+	}
+
+	// Translate algorithm back to string
+	algorithm := key.Algorithm.String()
+
+	// Return values of key
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"issuer":       key.Issuer,
+			"account_name": key.AccountName,
+			"period":       key.Period,
+			"algorithm":    algorithm,
+			"digits":       key.Digits,
+		},
+	}, nil
+}
+
+func (b *backend) pathKeyList(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	entries, err := req.Storage.List("key/")
+	if err != nil {
+		return nil, err
+	}
+
+	return logical.ListResponse(entries), nil
+}
+
+func (b *backend) pathKeyCreate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+	generate := data.Get("generate").(bool)
+	keyString := data.Get("key").(string)
+	issuer := data.Get("issuer").(string)
+	accountName := data.Get("account_name").(string)
+	period := data.Get("period").(int)
+	algorithm := data.Get("algorithm").(string)
+	digits := data.Get("digits").(int)
+	skew := data.Get("skew").(int)
+	qrSize := data.Get("qr_size").(int)
+	keySize := data.Get("key_size").(int)
+	inputURL := data.Get("url").(string)
```

Can we throw an error if both generate is true and an inputURL is also set? I think the same should happen if generate is true and keyString is set.